File: src/ingestion/chains/lidl.ts
/**
 * Lidl Chain Adapter
 *
 * Adapter for parsing Lidl retail chain price data files.
 * Lidl uses CSV format with comma delimiter and UTF-8 encoding.
 * Store resolution is based on filename (daily ZIP files -> fanout).
 * Handles multiple GTINs per SKU.
 */

import type {
  DiscoveredFile,
  NormalizedRow,
  NormalizedRowValidation,
  ParseResult,
} from '../core/types'
import type { CsvColumnMapping } from '../parsers/csv'
import { BaseCsvAdapter } from './base'
import { CHAIN_CONFIGS } from './config'

/**
 * Column mapping for Lidl CSV files (2026 format).
 * Maps Lidl's column names to NormalizedRow fields.
 * Lidl may include multiple GTINs separated by semicolon in the barcode field.
 *
 * Current columns (as of 2026):
 * NAZIV, ŠIFRA, NETO_KOLIČINA, JEDINICA_MJERE, MARKA, MALOPRODAJNA_CIJENA,
 * MPC_ZA_VRIJEME_POSEBNOG_OBLIKA_PRODAJE, NAJNIZA_CIJENA_U_POSLJ._30_DANA,
 * CIJENA_ZA_JEDINICU_MJERE, BARKOD, KATEGORIJA_PROIZVODA, Sidrena_cijena_na_...
 */
const LIDL_COLUMN_MAPPING: CsvColumnMapping = {
  externalId: 'ŠIFRA',
  name: 'NAZIV',
  category: 'KATEGORIJA_PROIZVODA',
  brand: 'MARKA',
  unit: 'JEDINICA_MJERE',
  unitQuantity: 'NETO_KOLIČINA',
  price: 'MALOPRODAJNA_CIJENA',
  discountPrice: 'MPC_ZA_VRIJEME_POSEBNOG_OBLIKA_PRODAJE',
  barcodes: 'BARKOD',
}

/**
 * Alternative column mapping for Lidl CSV files (legacy format).
 * Some older Lidl exports may use these column names.
 */
const LIDL_COLUMN_MAPPING_ALT: CsvColumnMapping = {
  externalId: 'Artikl',
  name: 'Naziv artikla',
  category: 'Kategorija',
  brand: 'Robna marka',
  unit: 'Jedinica mjere',
  unitQuantity: 'Količina',
  price: 'Cijena',
  discountPrice: 'Akcijska cijena',
  discountStart: 'Početak akcije',
  discountEnd: 'Završetak akcije',
  barcodes: 'GTIN',
}

/**
 * Lidl chain adapter implementation.
 * Extends BaseCsvAdapter with custom GTIN handling and discovery logic.
 *
 * Lidl portal: https://tvrtka.lidl.hr/cijene
 * Download URL pattern: https://tvrtka.lidl.hr/content/download/[DYNAMIC_ID]/fileupload/[FILENAME].zip
 * Filename format: Popis_cijena_po_trgovinama_na_dan_DD_MM_YYYY.zip
 */
export class LidlAdapter extends BaseCsvAdapter {
  /** Date to discover files for (YYYY-MM-DD format, set by CLI before discovery) */
  private discoveryDate: string | null = null

  constructor() {
    super({
      slug: 'lidl',
      name: 'Lidl',
      supportedTypes: ['csv', 'zip'],
      chainConfig: CHAIN_CONFIGS.lidl,
      columnMapping: LIDL_COLUMN_MAPPING,
      alternativeColumnMapping: LIDL_COLUMN_MAPPING_ALT,
      filenamePrefixPatterns: [
        /^Lidl[_-]?/i,
        /^Popis_cijena[_-]?/i,
        /^cjenik[_-]?/i,
        /^\d{4}[_-]\d{2}[_-]\d{2}[_-]?/, // Remove date prefix
      ],
      rateLimitConfig: {
        requestsPerSecond: 2,
        maxRetries: 3,
      },
    })
  }

  /**
   * Set the date to use for discovery.
   * @param date - Date in YYYY-MM-DD format
   */
  setDiscoveryDate(date: string): void {
    this.discoveryDate = date
  }

  /**
   * Convert date from YYYY-MM-DD to DD_MM_YYYY format (Lidl filename format).
   */
  private formatDateForFilename(date: string): string {
    const [year, month, day] = date.split('-')
    return `${day}_${month}_${year}`
  }

  /**
   * Extract date from Lidl filename (DD_MM_YYYY) to YYYY-MM-DD format.
   */
  private extractDateFromFilename(filename: string): string | null {
    // Pattern: Popis_cijena_po_trgovinama_na_dan_DD_MM_YYYY.zip
    const match = filename.match(/(\d{2})_(\d{2})_(\d{4})\.zip$/i)
    if (match) {
      const [, day, month, year] = match
      return `${year}-${month}-${day}`
    }
    return null
  }

  /**
   * Discover available Lidl price files.
   *
   * Lidl's portal uses dynamic download IDs that must be parsed from the HTML:
   * - URL format: https://tvrtka.lidl.hr/cijene
   * - Download links: /content/download/[ID]/fileupload/Popis_cijena_po_trgovinama_na_dan_DD_MM_YYYY.zip
   *
   * @returns Array of discovered files (filtered by date if setDiscoveryDate was called)
   */
  async discover(): Promise<DiscoveredFile[]> {
    const discoveredFiles: DiscoveredFile[] = []
    const seenUrls = new Set<string>()

    try {
      const response = await fetch(this.config.baseUrl, {
        headers: {
          'User-Agent': 'Mozilla/5.0 (compatible; PriceTracker/1.0)',
          'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
        },
      })

      if (!response.ok) {
        console.error(`Failed to fetch Lidl portal: ${response.status} ${response.statusText}`)
        return []
      }

      const html = await response.text()

      // Extract download links matching Lidl's URL pattern
      // Pattern: href="(https://tvrtka.lidl.hr/content/download/\d+/fileupload/Popis_cijena[^"]+\.zip)"
      const downloadPattern = /href=["'](https:\/\/tvrtka\.lidl\.hr\/content\/download\/\d+\/fileupload\/([^"']+\.zip))["']/gi

      let match: RegExpExecArray | null
      while ((match = downloadPattern.exec(html)) !== null) {
        const fileUrl = match[1]
        const filename = match[2]

        // Skip duplicates
        if (seenUrls.has(fileUrl)) {
          continue
        }
        seenUrls.add(fileUrl)

        // Extract date from filename
        const fileDate = this.extractDateFromFilename(filename)

        // If discoveryDate is set, filter to only that date
        if (this.discoveryDate && fileDate !== this.discoveryDate) {
          continue
        }

        discoveredFiles.push({
          url: fileUrl,
          filename,
          type: 'zip',
          size: null,
          lastModified: fileDate ? new Date(fileDate) : null,
          metadata: {
            source: 'lidl_portal',
            discoveredAt: new Date().toISOString(),
            ...(fileDate && { portalDate: fileDate }),
          },
        })
      }

      // If using specific date and no exact match found, try relative URL pattern
      if (this.discoveryDate && discoveredFiles.length === 0) {
        // Try alternative pattern for relative URLs
        const relativePattern = /href=["'](\/content\/download\/\d+\/fileupload\/([^"']+\.zip))["']/gi

        while ((match = relativePattern.exec(html)) !== null) {
          const href = match[1]
          const filename = match[2]
          const fileUrl = new URL(href, this.config.baseUrl).toString()

          if (seenUrls.has(fileUrl)) {
            continue
          }
          seenUrls.add(fileUrl)

          const fileDate = this.extractDateFromFilename(filename)

          if (this.discoveryDate && fileDate !== this.discoveryDate) {
            continue
          }

          discoveredFiles.push({
            url: fileUrl,
            filename,
            type: 'zip',
            size: null,
            lastModified: fileDate ? new Date(fileDate) : null,
            metadata: {
              source: 'lidl_portal',
              discoveredAt: new Date().toISOString(),
              ...(fileDate && { portalDate: fileDate }),
            },
          })
        }
      }

      return discoveredFiles
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error)
      console.error(`Error discovering Lidl files: ${errorMessage}`)
      return []
    }
  }

  /**
   * Override base extractFilenameFromUrl to use 'unknown.zip' as default.
   */
  protected override extractFilenameFromUrl(url: string): string {
    try {
      const urlObj = new URL(url)
      const pathname = urlObj.pathname
      const filename = pathname.split('/').pop() || 'unknown.zip'
      return filename.split('?')[0]
    } catch {
      return 'unknown.zip'
    }
  }

  /**
   * Post-process parse result to handle multiple GTINs.
   * Lidl may list multiple GTINs separated by semicolon or pipe.
   */
  protected postprocessResult(result: ParseResult): ParseResult {
    result.rows = result.rows.map((row) => this.normalizeMultipleGtins(row))
    return result
  }

  /**
   * Normalize multiple GTINs in a row.
   * Lidl may list multiple GTINs separated by semicolon or pipe.
   */
  private normalizeMultipleGtins(row: NormalizedRow): NormalizedRow {
    if (row.barcodes.length === 1) {
      // Check if single barcode contains multiple GTINs
      const barcode = row.barcodes[0]
      if (barcode.includes(';') || barcode.includes('|')) {
        const gtins = barcode
          .split(/[;|]/)
          .map((g) => g.trim())
          .filter((g) => g.length > 0)
        return { ...row, barcodes: gtins }
      }
    }
    return row
  }

  /**
   * Extract store identifier string from filename.
   * Lidl has special patterns for store identification.
   */
  protected extractStoreIdentifierFromFilename(filename: string): string {
    // Remove file extension
    const baseName = filename.replace(/\.(csv|CSV)$/, '')

    // Try to extract store ID from various Lidl filename patterns
    // Pattern 1: Lidl_DATE_STOREID (e.g., "Lidl_2024-01-15_42")
    const dateStoreMatch = baseName.match(/^Lidl[_-]?\d{4}[_-]\d{2}[_-]\d{2}[_-](.+)$/i)
    if (dateStoreMatch) {
      return dateStoreMatch[1]
    }

    // Pattern 2: Lidl_Poslovnica_LOCATION (e.g., "Lidl_Poslovnica_Zagreb_Ilica_123")
    const locationMatch = baseName.match(/^Lidl[_-]?Poslovnica[_-]?(.+)$/i)
    if (locationMatch) {
      return locationMatch[1]
    }

    // Pattern 3: Just Lidl_STOREID (e.g., "Lidl_42")
    const simpleMatch = baseName.match(/^Lidl[_-]?(\d+)$/i)
    if (simpleMatch) {
      return simpleMatch[1]
    }

    // Fall back to base class implementation
    return super.extractStoreIdentifierFromFilename(filename)
  }

  /**
   * Validate a normalized row according to Lidl-specific rules.
   * Uses stricter GTIN validation.
   */
  validateRow(row: NormalizedRow): NormalizedRowValidation {
    // Get base validation from parent
    const baseValidation = super.validateRow(row)
    const warnings = [...baseValidation.warnings]

    // Replace barcode warnings with Lidl-specific GTIN validation
    const gtinWarnings = warnings.filter(w => !w.includes('Invalid barcode format'))

    // GTIN validation - Lidl uses EAN-13 and EAN-8
    for (const barcode of row.barcodes) {
      if (!/^\d{8}$|^\d{13}$|^\d{14}$/.test(barcode)) {
        gtinWarnings.push(`Invalid GTIN format: ${barcode} (expected EAN-8, EAN-13, or GTIN-14)`)
      }
    }

    // Lidl products should typically have at least one GTIN
    if (row.barcodes.length === 0) {
      gtinWarnings.push('No GTIN/barcode found for product')
    }

    return {
      isValid: baseValidation.isValid,
      errors: baseValidation.errors,
      warnings: gtinWarnings,
    }
  }
}

/**
 * Create a Lidl adapter instance.
 */
export function createLidlAdapter(): LidlAdapter {
  return new LidlAdapter()
}

File: services/price-service/internal/adapters/chains/lidl.go
package chains

import (
	"context"
	"fmt"
	"io"
	"regexp"
	"strings"
	"time"

	"github.com/rs/zerolog/log"
	"github.com/kosarica/price-service/internal/adapters/base"
	"github.com/kosarica/price-service/internal/adapters/config"
	zipexpand "github.com/kosarica/price-service/internal/ingestion/zip"
	"github.com/kosarica/price-service/internal/parsers/csv"
	"github.com/kosarica/price-service/internal/types"
)

// lidlColumnMapping is the primary column mapping for Lidl CSV files (2026 format)
// Maps Lidl's Croatian column names to NormalizedRow fields
var lidlColumnMapping = csv.CsvColumnMapping{
	ExternalID:     types.StringPtr("ŠIFRA"),
	Name:           "NAZIV",
	Category:       types.StringPtr("KATEGORIJA_PROIZVODA"),
	Brand:          types.StringPtr("MARKA"),
	Unit:           types.StringPtr("JEDINICA_MJERE"),
	UnitQuantity:   types.StringPtr("NETO_KOLIČINA"),
	Price:          "MALOPRODAJNA_CIJENA",
	DiscountPrice:  types.StringPtr("MPC_ZA_VRIJEME_POSEBNOG_OBLIKA_PRODAJE"),
	Barcodes:       types.StringPtr("BARKOD"),
	UnitPrice:      types.StringPtr("CIJENA_ZA_JEDINICU_MJERE"),
	LowestPrice30d: types.StringPtr("NAJNIZA_CIJENA_U_POSLJ._30_DANA"),
	AnchorPrice:    types.StringPtr("Sidrena_cijena_na_dan"),
}

// lidlColumnMappingAlt is the alternative column mapping for Lidl CSV files (legacy format)
var lidlColumnMappingAlt = csv.CsvColumnMapping{
	ExternalID:    types.StringPtr("Artikl"),
	Name:          "Naziv artikla",
	Category:      types.StringPtr("Kategorija"),
	Brand:         types.StringPtr("Robna marka"),
	Unit:          types.StringPtr("Jedinica mjere"),
	UnitQuantity:  types.StringPtr("Količina"),
	Price:         "Cijena",
	DiscountPrice: types.StringPtr("Akcijska cijena"),
	DiscountStart: types.StringPtr("Početak akcije"),
	DiscountEnd:   types.StringPtr("Završetak akcije"),
	Barcodes:      types.StringPtr("GTIN"),
}

// LidlAdapter is the chain adapter for Lidl retail chain
type LidlAdapter struct {
	*base.BaseCsvAdapter
	discoveryDate string // Optional date filter for discovery (YYYY-MM-DD)
}

// NewLidlAdapter creates a new Lidl adapter
func NewLidlAdapter() (*LidlAdapter, error) {
	chainConfig := config.ChainConfigs[config.ChainLidl]

	adapterConfig := base.CsvAdapterConfig{
		BaseAdapterConfig: base.BaseAdapterConfig{
			Slug:           string(config.ChainLidl),
			Name:           chainConfig.Name,
			SupportedTypes: []types.FileType{types.FileTypeCSV, types.FileTypeZIP},
			ChainConfig:    chainConfig,
			FilenamePrefixPatterns: []string{
				`(?i)^Lidl[_-]?`,
				`(?i)^Popis_cijena[_-]?`,
				`(?i)^cjenik[_-]?`,
				`^\d{4}[_-]\d{2}[_-]\d{2}[_-]?`, // Remove date prefix
			},
			FileExtensionPattern: regexp.MustCompile(`\.(csv|CSV|zip|ZIP)$`),
		},
		ColumnMapping:            lidlColumnMapping,
		AlternativeColumnMapping: &lidlColumnMappingAlt,
	}

	baseAdapter, err := base.NewBaseCsvAdapter(adapterConfig)
	if err != nil {
		return nil, fmt.Errorf("failed to create base CSV adapter: %w", err)
	}

	return &LidlAdapter{
		BaseCsvAdapter: baseAdapter,
	}, nil
}

// SetDiscoveryDate sets the date to use for discovery filtering
func (a *LidlAdapter) SetDiscoveryDate(date string) {
	a.discoveryDate = date
}

// Discover discovers available price files from Lidl portal
// Lidl uses dynamic download IDs that must be parsed from HTML:
// URL pattern: https://tvrtka.lidl.hr/content/download/[ID]/fileupload/Popis_cijena_po_trgovinama_na_dan_DD_MM_YYYY.zip
func (a *LidlAdapter) Discover(targetDate string) ([]types.DiscoveredFile, error) {
	discoveredFiles := make([]types.DiscoveredFile, 0)
	seenURLs := make(map[string]bool)

	// Use provided targetDate or the pre-set discoveryDate
	filterDate := targetDate
	if filterDate == "" {
		filterDate = a.discoveryDate
	}

	log.Debug().Str("url", a.BaseURL()).Msg("Fetching Lidl portal")

	resp, err := a.HTTPClient().Get(a.BaseURL())
	if err != nil {
		log.Error().Err(err).Msg("Failed to fetch Lidl portal")
		return nil, fmt.Errorf("failed to fetch Lidl portal: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		log.Error().Int("status_code", resp.StatusCode).Msg("Lidl portal returned error status")
		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
	}

	bodyBytes, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response body: %w", err)
	}

	html := string(bodyBytes)

	// Extract download links matching Lidl's URL pattern
	// Pattern: href="(https://tvrtka.lidl.hr/content/download/\d+/fileupload/[^"]+\.zip)"
	downloadPattern := regexp.MustCompile(`href=["'](https://tvrtka\.lidl\.hr/content/download/\d+/fileupload/([^"']+\.zip))["']`)

	matches := downloadPattern.FindAllStringSubmatch(html, -1)
	for _, match := range matches {
		if len(match) < 3 {
			continue
		}

		fileURL := match[1]
		filename := match[2]

		// Skip duplicates
		if seenURLs[fileURL] {
			continue
		}
		seenURLs[fileURL] = true

		// Extract date from filename
		fileDate := a.extractDateFromFilename(filename)

		// Filter by date if specified
		if filterDate != "" && fileDate != filterDate {
			continue
		}

		var lastModified *time.Time
		if fileDate != "" {
			if t, err := time.Parse("2006-01-02", fileDate); err == nil {
				lastModified = &t
			}
		}

		discoveredFiles = append(discoveredFiles, types.DiscoveredFile{
			URL:          fileURL,
			Filename:     filename,
			Type:         types.FileTypeZIP,
			Size:         nil,
			LastModified: lastModified,
			Metadata: map[string]string{
				"source":       "lidl_portal",
				"discoveredAt": time.Now().Format(time.RFC3339),
				"portalDate":   fileDate,
			},
		})
	}

	// If no files found with absolute URLs, try relative URL pattern
	if len(discoveredFiles) == 0 {
		relativePattern := regexp.MustCompile(`href=["'](/content/download/\d+/fileupload/([^"']+\.zip))["']`)

		matches = relativePattern.FindAllStringSubmatch(html, -1)
		for _, match := range matches {
			if len(match) < 3 {
				continue
			}

			href := match[1]
			filename := match[2]

			// Build full URL
			fileURL := "https://tvrtka.lidl.hr" + href

			if seenURLs[fileURL] {
				continue
			}
			seenURLs[fileURL] = true

			fileDate := a.extractDateFromFilename(filename)

			if filterDate != "" && fileDate != filterDate {
				continue
			}

			var lastModified *time.Time
			if fileDate != "" {
				if t, err := time.Parse("2006-01-02", fileDate); err == nil {
					lastModified = &t
				}
			}

			discoveredFiles = append(discoveredFiles, types.DiscoveredFile{
				URL:          fileURL,
				Filename:     filename,
				Type:         types.FileTypeZIP,
				Size:         nil,
				LastModified: lastModified,
				Metadata: map[string]string{
					"source":       "lidl_portal",
					"discoveredAt": time.Now().Format(time.RFC3339),
					"portalDate":   fileDate,
				},
			})
		}
	}

	log.Debug().Int("file_count", len(discoveredFiles)).Msg("Discovered files from Lidl portal")
	return discoveredFiles, nil
}

// extractDateFromFilename extracts date from Lidl filename (DD_MM_YYYY) to YYYY-MM-DD format
func (a *LidlAdapter) extractDateFromFilename(filename string) string {
	// Pattern: Popis_cijena_po_trgovinama_na_dan_DD_MM_YYYY.zip
	match := regexp.MustCompile(`(\d{2})_(\d{2})_(\d{4})\.zip$`).FindStringSubmatch(filename)
	if len(match) >= 4 {
		day := match[1]
		month := match[2]
		year := match[3]
		return fmt.Sprintf("%s-%s-%s", year, month, day)
	}
	return ""
}

// ExpandZIP expands a ZIP file and returns the extracted CSV files
func (a *LidlAdapter) ExpandZIP(ctx context.Context, content []byte, filename string) ([]zipexpand.ExpandedFile, error) {
	expanded, err := zipexpand.ExpandInMemory(content, filename)
	if err != nil {
		return nil, fmt.Errorf("failed to expand ZIP: %w", err)
	}

	// Filter to only CSV files
	csvFiles := make([]zipexpand.ExpandedFile, 0, len(expanded))
	for _, file := range expanded {
		if file.Type == types.FileTypeCSV {
			csvFiles = append(csvFiles, file)
		}
	}

	log.Debug().Int("csv_count", len(csvFiles)).Str("filename", filename).Msg("Expanded CSV files from ZIP")
	return csvFiles, nil
}

// Parse parses CSV content with multiple GTIN handling
func (a *LidlAdapter) Parse(content []byte, filename string, options *types.ParseOptions) (*types.ParseResult, error) {
	// Use base adapter's Parse method
	result, err := a.BaseCsvAdapter.Parse(content, filename, options)
	if err != nil {
		return nil, err
	}

	// Post-process to handle multiple GTINs
	result = a.postprocessMultipleGTINs(result)

	return result, nil
}

// postprocessMultipleGTINs handles Lidl's multiple GTINs in barcode field
// Lidl may list multiple GTINs separated by semicolon or pipe
func (a *LidlAdapter) postprocessMultipleGTINs(result *types.ParseResult) *types.ParseResult {
	for i := range result.Rows {
		row := &result.Rows[i]
		if len(row.Barcodes) == 1 {
			barcode := row.Barcodes[0]
			// Check if single barcode contains multiple GTINs
			if strings.Contains(barcode, ";") || strings.Contains(barcode, "|") {
				gtins := splitGTINs(barcode)
				row.Barcodes = gtins
			}
		}
	}
	return result
}

// splitGTINs splits a barcode string containing multiple GTINs
func splitGTINs(barcode string) []string {
	// Split on semicolon or pipe
	parts := regexp.MustCompile(`[;|]`).Split(barcode, -1)

	gtins := make([]string, 0, len(parts))
	for _, part := range parts {
		gtin := strings.TrimSpace(part)
		if gtin != "" {
			gtins = append(gtins, gtin)
		}
	}

	return gtins
}

// ExtractStoreIdentifierFromFilename extracts store identifier from Lidl filename
// Lidl has special patterns for store identification
func (a *LidlAdapter) ExtractStoreIdentifierFromFilename(filename string) string {
	// Remove file extension
	baseName := regexp.MustCompile(`\.(csv|CSV)$`).ReplaceAllString(filename, "")

	// Pattern 1: Lidl_DATE_STOREID (e.g., "Lidl_2024-01-15_42")
	dateStoreMatch := regexp.MustCompile(`(?i)^Lidl[_-]?\d{4}[_-]\d{2}[_-]\d{2}[_-](.+)$`).FindStringSubmatch(baseName)
	if len(dateStoreMatch) >= 2 {
		return dateStoreMatch[1]
	}

	// Pattern 2: Lidl_Poslovnica_LOCATION (e.g., "Lidl_Poslovnica_Zagreb_Ilica_123")
	locationMatch := regexp.MustCompile(`(?i)^Lidl[_-]?Poslovnica[_-]?(.+)$`).FindStringSubmatch(baseName)
	if len(locationMatch) >= 2 {
		return locationMatch[1]
	}

	// Pattern 3: Just Lidl_STOREID (e.g., "Lidl_42")
	simpleMatch := regexp.MustCompile(`(?i)^Lidl[_-]?(\d+)$`).FindStringSubmatch(baseName)
	if len(simpleMatch) >= 2 {
		return simpleMatch[1]
	}

	// Pattern 4: Supermarket 265_Address_... format
	// Example: "Supermarket 265_Ulica Franje Glada_13_40323_Prelog_1_16.12.2025_7.15h.csv"
	parts := strings.Split(baseName, "_")
	if len(parts) > 0 {
		firstPart := strings.Split(parts[0], " ")
		if len(firstPart) >= 2 {
			// Return store type + ID (e.g., "Supermarket 265")
			return parts[0]
		}
	}

	// Fall back: use base adapter method via creating a dummy file
	dummyFile := types.DiscoveredFile{Filename: filename}
	if id := a.ExtractStoreIdentifier(dummyFile); id != nil {
		return id.Value
	}

	return ""
}

// ValidateRow validates a normalized row with Lidl-specific GTIN validation
func (a *LidlAdapter) ValidateRow(row types.NormalizedRow) types.NormalizedRowValidation {
	// Get base validation
	baseValidation := a.BaseCsvAdapter.ValidateRow(row)

	// Build warnings list without generic barcode warnings
	warnings := make([]string, 0)
	for _, w := range baseValidation.Warnings {
		if !strings.Contains(w, "Invalid barcode format") {
			warnings = append(warnings, w)
		}
	}

	// Add Lidl-specific GTIN validation
	for _, barcode := range row.Barcodes {
		if !isValidGTIN(barcode) {
			warnings = append(warnings, fmt.Sprintf("Invalid GTIN format: %s (expected EAN-8, EAN-13, or GTIN-14)", barcode))
		}
	}

	// Lidl products should typically have at least one GTIN
	if len(row.Barcodes) == 0 {
		warnings = append(warnings, "No GTIN/barcode found for product")
	}

	return types.NormalizedRowValidation{
		IsValid:  baseValidation.IsValid,
		Errors:   baseValidation.Errors,
		Warnings: warnings,
	}
}

// isValidGTIN checks if a barcode is a valid GTIN (EAN-8, EAN-13, or GTIN-14)
func isValidGTIN(barcode string) bool {
	// Must be 8, 13, or 14 digits
	if len(barcode) != 8 && len(barcode) != 13 && len(barcode) != 14 {
		return false
	}

	// Must be all digits
	for _, c := range barcode {
		if c < '0' || c > '9' {
			return false
		}
	}

	return true
}

// ExtractStoreMetadata extracts store metadata from Lidl filename for auto-registration
// Parses filename pattern: {type} {storeId}_{address}_{number}_{postal}_{city}_{ver}_{date}_{time}.csv
// Example: Supermarket 265_Ulica Franje Glada_13_40323_Prelog_1_16.12.2025_7.15h.csv
func (a *LidlAdapter) ExtractStoreMetadata(file types.DiscoveredFile) *types.StoreMetadata {
	// Remove extension
	baseName := regexp.MustCompile(`\.(csv|CSV)$`).ReplaceAllString(file.Filename, "")

	// Split by underscore
	parts := strings.Split(baseName, "_")
	if len(parts) < 6 {
		// Fall back to base implementation
		storeID := a.ExtractStoreIdentifierFromFilename(file.Filename)
		return &types.StoreMetadata{
			Name: fmt.Sprintf("Lidl %s", storeID),
		}
	}

	// First part has space: "Supermarket 265"
	firstPart := strings.Split(parts[0], " ")
	storeType := ""
	if len(firstPart) >= 1 {
		storeType = firstPart[0]
	}

	// Find postal code (5 digits) to anchor the structure
	postalIdx := -1
	for i := 1; i < len(parts)-2; i++ {
		if regexp.MustCompile(`^\d{5}$`).MatchString(parts[i]) {
			postalIdx = i
			break
		}
	}

	if postalIdx == -1 {
		storeID := a.ExtractStoreIdentifierFromFilename(file.Filename)
		return &types.StoreMetadata{
			Name: fmt.Sprintf("Lidl %s", storeID),
		}
	}

	// Address is from index 1 to postalIdx-1
	addressParts := parts[1:postalIdx]
	address := strings.Join(addressParts, " ")
	postalCode := parts[postalIdx]
	city := ""
	if postalIdx+1 < len(parts) {
		city = parts[postalIdx+1]
	}

	return &types.StoreMetadata{
		Name:       fmt.Sprintf("Lidl %s", titleCase(city)),
		Address:    titleCase(address),
		City:       titleCase(city),
		PostalCode: postalCode,
		StoreType:  titleCase(storeType),
	}
}

// titleCase converts string to title case
func titleCase(s string) string {
	words := strings.Fields(strings.ToLower(s))
	for i, word := range words {
		if len(word) > 0 {
			words[i] = strings.ToUpper(string(word[0])) + word[1:]
		}
	}
	return strings.Join(words, " ")
}

// UsesZIP returns true as Lidl uses ZIP files
func (a *LidlAdapter) UsesZIP() bool {
	return true
}
